### ISSUE TITLE
Question about Gaussian rotations

### ISSUE THREAD

**Issue by RoboPerception** [hustvl/GaussTR#40]
https://github.com/hustvl/GaussTR/blob/main/gausstr/models/gausstr_head.py#L152C9-L152C18
It seems that all the Gaussian rotations are fixed by cam2ego, and the model does not learn to predict Gaussian rotations. So why not let the model learn Gaussian rotations dynamically?

**Comment by RoboPerception** [comment]
### 

> By initializing the Gaussian rotations orthogonal to the camera plane, it creates a consistent learning target across different camera views.
> 
> I think it's feasible to predict a Δrotation upon this initialization, but predicting the quaternion requires extra constraints, and predicting the scales with fixed rotations can already handle the anisotropy to some extent, so we do not introduce the rotation prediction in the current version.

Thank you for your explanation. I have a follow-up question regarding the use of rendered depth and ground truth depth. In GaussTR, depth maps from Metric3D are used as ground truth for depth supervision. However, there is an inherent gap between the depth estimated by Metric3D and the actual real-world depth. Given this discrepancy, why is it still effective to use Metric3D depth as ground truth? Additionally, during training, the model predicts an offset from the Metric3D depth to determine the center of the Gaussian kernel, which is then used to render a depth map and compute the loss against the Metric3D depth. Doesn’t this raise a question: if the model is supervised to match the Metric3D depth, could it simply learn to place the Gaussian kernel centers directly at the Metric3D depth values without learning any meaningful offset?

**Comment by npurson** [comment]
> > By initializing the Gaussian rotations orthogonal to the camera plane, it creates a consistent learning target across different camera views.
> > I think it's feasible to predict a Δrotation upon this initialization, but predicting the quaternion requires extra constraints, and predicting the scales with fixed rotations can already handle the anisotropy to some extent, so we do not introduce the rotation prediction in the current version.
> 
> Thank you for your explanation. I have a follow-up question regarding the use of rendered depth and ground truth depth. In GaussTR, depth maps from Metric3D are used as ground truth for depth supervision. However, there is an inherent gap between the depth estimated by Metric3D and the actual real-world depth. Given this discrepancy, why is it still effective to use Metric3D depth as ground truth? Additionally, during training, the model predicts an offset from the Metric3D depth to determine the center of the Gaussian kernel, which is then used to render a depth map and compute the loss against the Metric3D depth. Doesn’t this raise a question: if the model is supervised to match the Metric3D depth, could it simply learn to place the Gaussian kernel centers directly at the Metric3D depth values without learning any meaningful offset?

Excellent questions. I appreciate you taking the time to think through the details of our approach. Here’s our explanation:

1. Indeed, using gt depth would likely produce the best results. However, requirements on depth-paired training data would significantly limit the data scalability of our SSL approach. Our goal is to develop a 3D SSL method that can be aligned with a variety of 2D VFMs across diverse domains, to leverage their progress and alleviate the challenge of 3D representation learning. Metric3D provides a good compromise under the circumstance.

2. Yes, I agree if we were to use pixel-aligned Gaussians (one Gaussian per pixel) and initialize their positions with depth, it can hardly learn the reasonable position offset. However, our method use sparse Gaussian representations, with only 300 spanning the entire images. This **sparsity** is key. A single Gaussian is responsible for representing not just a single point on a surface, but a larger neighborhood and contributing to the overall shape of an object. Consider a car in an image as an example. If we initialize a Gaussian **on the surface** of the car using depth, but the model is expected to learn to place the Gaussian at **the center inside the car's volume**. Beside, the Gaussians have to satisfy the multi-view consistency within the overlapped regions between views. I would recommend taking a look at the visualization videos in our project page, which indicate this behavior as well.

I hope this addresses your questions.

**Comment by npurson** [comment]
> > > > > > By initializing the Gaussian rotations orthogonal to the camera plane, it creates a consistent learning target across different camera views.
> > > > > > I think it's feasible to predict a Δrotation upon this initialization, but predicting the quaternion requires extra constraints, and predicting the scales with fixed rotations can already handle the anisotropy to some extent, so we do not introduce the rotation prediction in the current version.
> > > > > 
> > > > > 
> > > > > Thank you for your explanation. I have a follow-up question regarding the use of rendered depth and ground truth depth. In GaussTR, depth maps from Metric3D are used as ground truth for depth supervision. However, there is an inherent gap between the depth estimated by Metric3D and the actual real-world depth. Given this discrepancy, why is it still effective to use Metric3D depth as ground truth? Additionally, during training, the model predicts an offset from the Metric3D depth to determine the center of the Gaussian kernel, which is then used to render a depth map and compute the loss against the Metric3D depth. Doesn’t this raise a question: if the model is supervised to match the Metric3D depth, could it simply learn to place the Gaussian kernel centers directly at the Metric3D depth values without learning any meaningful offset?
> > > > 
> > > > 
> > > > Excellent questions. I appreciate you taking the time to think through the details of our approach. Here’s our explanation:
> > > > 
> > > > 1. Indeed, using gt depth would likely produce the best results. However, requirements on depth-paired training data would significantly limit the data scalability of our SSL approach. Our goal is to develop a 3D SSL method that can be aligned with a variety of 2D VFMs across diverse domains, to leverage their progress and alleviate the challenge of 3D representation learning. Metric3D provides a good compromise under the circumstance.
> > > > 2. Yes, I agree if we were to use pixel-aligned Gaussians (one Gaussian per pixel) and initialize their positions with depth, it can hardly learn the reasonable position offset. However, our method use sparse Gaussian representations, with only 300 spanning the entire images. This **sparsity** is key. A single Gaussian is responsible for representing not just a single point on a surface, but a larger neighborhood and contributing to the overall shape of an object. Consider a car in an image as an example. If we initialize a Gaussian **on the surface** of the car using depth, but the model is expected to learn to place the Gaussian at **the center inside the car's volume**. Beside, the Gaussians have to satisfy the **multi-view consistency** within the overlapped regions between views. I would recommend taking a look at the visualization videos in our project page, which indicate this behavior as well.
> > > > 
> > > > I hope this addresses your questions.
> > > 
> > > 
> > > Thank you. I got it. But I'm curious about the difference between the depth generated by Metric3D and the ground truth depth on nuScenes. Have you conducted any tests on this?
> > 
> > 
> > We haven't conducted such experiments. BTW, I remind that the gt depth provided by nuScenes is sparse as it is obtained by LiDAR. It requires extra processing to densify it. This is also one of reasons why we don't use it.
> 
> Thanks for your detailed explanation. I'm sorry to bother you again, I meet some trouble in Gaussian voxelization. I conducted an experiment on KITTI. The voxels output by Gaussian voxelization are very sparse, especially in the distance. Then I adjusted the [parameters](https://github.com/hustvl/GaussTR/blob/main/gausstr/models/gaussian_voxelizer.py#L37), and the voxels became denser, but also thicker, and the height of the ground area increased. Could you explain how to adjust the voxelization parameters?
> 
> <img alt="Image" width="478" height="409" src="https://private-user-images.githubusercontent.com/185015151/471027981-0849509f-374a-416f-80b7-15cf3dce0ae0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM0OTczODMsIm5iZiI6MTc1MzQ5NzA4MywicGF0aCI6Ii8xODUwMTUxNTEvNDcxMDI3OTgxLTA4NDk1MDlmLTM3NGEtNDE2Zi04MGI3LTE1Y2YzZGNlMGFlMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcyNlQwMjMxMjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZGI4MmNjOWNmMTlkOGMzNGVhMjg5NDc4ZGQzMDRmMTdlOTc5MzFlMTg2NzgzNzMwMzMyMTMxOGYyNjAxMGQ4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.o6ykwNn5abRVLOoLlweq3fWLCCgQpEVdMe-MD1bI5p4"> <img alt="Image" width="478" height="409" src="https://private-user-images.githubusercontent.com/185015151/471027996-b6cff308-0881-4640-ac3b-e67e6a388841.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM0OTczODMsIm5iZiI6MTc1MzQ5NzA4MywicGF0aCI6Ii8xODUwMTUxNTEvNDcxMDI3OTk2LWI2Y2ZmMzA4LTA4ODEtNDY0MC1hYzNiLWU2N2U2YTM4ODg0MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcyNlQwMjMxMjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMGNkN2RhM2RkYjViMDhmOGIwZmUxYTk5MDhhNThkN2U1YjM5ZjFiMTRkYjI3ZTBkZWFhNWY0ZmI1MmFkNWRjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Wn7hTFgdAaXoHWWJFh16yliU6t0FR5PVK3qx2Y-3aKQ">

Do the voxelized Gaussians have a relatively correct position? If then, can you visualize the 2D rendered depth, and the target depth for supervision? Can you evaluate its IoU and mIoU metrics? You may also tune the density threshold:

https://github.com/hustvl/GaussTR/blob/247c88ed22794cc186f53c379bc7a110a74efb3b/gausstr/models/gausstr_head.py#L170C33-L170C70

My primary hypothesis is that the model struggles to optimize distant areas using solely 2D supervision. This problem is likely exacerbated on the KITTI dataset due to its longer perception range (51.2m) compared to nuScenes (40m), making geometric inaccuracies in the far-field more pronounced.

I assume a potential solution could be to leverage temporal information. By warping the predicted Gaussians to subsequent frames—where distant objects appear closer—we can apply depth supervision from a more advantageous viewpoint. This approach could significantly improve the geometric consistency and accuracy of the Gaussians, especially for distant regions.

**Comment by npurson** [comment]
Then this observation appears to be an inherent limitation of using monocular images for long-distance estimation. Leveraging temporal information might be a promising approach to address this.

**Comment by RoboPerception** [comment]
> Then this observation appears to be an inherent limitation of using monocular images for long-distance estimation. Leveraging temporal information might be a promising approach to address this.

Thank you. I will try it later.

**Comment by RoboPerception** [comment]
> Then this observation appears to be an inherent limitation of using monocular images for long-distance estimation. Leveraging temporal information might be a promising approach to address this.

Sorry to bother you again. I have some new confusions. I noticed that the target feature map is 16x down sample of input images, so why do you still use FeatUp to upsample the feature map.

**Comment by npurson** [comment]
> > > > Then this observation appears to be an inherent limitation of using monocular images for long-distance estimation. Leveraging temporal information might be a promising approach to address this.
> > > 
> > > 
> > > Sorry to bother you again. I have some new confusions. I noticed that the target feature map is 16x down sample of input images, so why do you still use FeatUp to upsample the feature map.
> > 
> > 
> > FeatUp not only improves feature resolution but also refines their spatial locality, compensating for a limitation in the original CLIP features. We use the downsampled features to maintain compatibility with other foundation models, such as various CLIP variants and Talk2DINO. As shown in Table 3 of our paper, augmenting the standard CLIP features with FeatUp is effective on its own, yielding a performance gain of approximately 1 mIoU.
> > Besides, while using the high-resolution features does provide a slight performance improvement, it introduces significant computational overhead.
> 
> I found some points that may cause the experimental results to be too sparse. I did not use feature maps to supervise the rendering results, and I only used depth maps for supervision. Will this lead to sparse reconstruction results? However, I found that in your experimental results of Table 2, using only the depth map for supervision can also achieve relatively good reconstruction results, so I don’t know whether the problem lies in using only the depth map for supervision.
> 
> <img alt="Image" width="478" height="409" src="https://private-user-images.githubusercontent.com/185015151/471884449-5b4fcd8e-e2fb-4fb0-ac8f-dfdd0ca11add.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM3ODEyOTQsIm5iZiI6MTc1Mzc4MDk5NCwicGF0aCI6Ii8xODUwMTUxNTEvNDcxODg0NDQ5LTViNGZjZDhlLWUyZmItNGZiMC1hYzhmLWRmZGQwY2ExMWFkZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcyOVQwOTIzMTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNTA5YjQ3YWFhMDM5NmMyNWE5NWEyNGNjYzlmM2Y3NGYwYmI4MmNjNzZhNDc5ZTk0NDI4NGQ0OTI4NzU2N2E2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.WDcxEPi5lGAGZRtcLHRQu50ZF-zz68NKpF8CM-bh1yQ">

I suspect the sparse results at a distance are not primarily caused by the lack of feature supervision. It's more likely that the model already achieves a satisfactory rendered depth with sparse predictions in those areas, leading to a lack of further optimization.

I just think of another potential solution could be to increase the loss weight for pixels at a greater distance. As seen in your visualization, the details in distant areas are relatively low, yet the voxelized visualization shows many redundant queries at the image margins. Since distant pixels occupy a smaller portion of the image due to perspective, weighting them more heavily might improve the results.

Besides, if you want to verify whether feature supervision is a cause, but worry about the costly FeatUp pre-processing, you could try the released Talk2DINO model, which doesn't require pre-processing.

**Comment by npurson** [comment]
> Thank you for your advice. I have verified that part of the reason why the previous results were too sparse was that feature supervision was not added, but I was confused about the code for calculating the depth loss, why do you set the distant areas in the depth map to 0.001?
https://github.com/hustvl/GaussTR/blob/main/gausstr/models/gausstr_head.py#L201

The goal is to only predict Gaussians to fit the geometry within self.depth_limit. For areas outside this range, if no Gaussians are present, the rendered depth would be zero, not the actual depth representing a very distant point. This is illustrated in Figure 4 of your paper, where the distant area is red (representing zero) in the "Rendered Depth" image, as opposed to the blue in the original "Depth" image.

To handle this, these distant areas should be set to 0. However, the SILog loss masks out zero values and ignores them during training. Therefore, by setting these areas to a very small value like 0.001, we expect that the network learns to render a depth of 0 for these regions.

**Comment by RoboPerception** [comment]
> > Thank you for your advice. I have verified that part of the reason why the previous results were too sparse was that feature supervision was not added, but I was confused about the code for calculating the depth loss, why do you set the distant areas in the depth map to 0.001?
> > https://github.com/hustvl/GaussTR/blob/main/gausstr/models/gausstr_head.py#L201
> 
> The goal is to only predict Gaussians to fit the geometry within self.depth_limit. For areas outside this range, if no Gaussians are present, the rendered depth would be zero, not the actual depth representing a very distant point. This is illustrated in Figure 4 of your paper, where the distant area is red (representing zero) in the "Rendered Depth" image, as opposed to the blue in the original "Depth" image.
> 
> To handle this, these distant areas should be set to 0. However, the SILog loss masks out zero values and ignores them during training. Therefore, by setting these areas to a very small value like 0.001, we expect that the network learns to render a depth of 0 for these regions.

Thank you, all my concerns are solved.
